go-controller/pkg/ovn/egressgw.go

```diff
@@ -585,48 +571,272 @@ func (oc *Controller) extSwitchPrefix(nodeName string) (string, error) {
 	return "", nil
 }
 
-// cleanECMPRoutes clean legacy ecmp routes when switching from
-// single bridge to default bridge plus second bridge used for external
-// gateway routes.
-func (oc *Controller) cleanECMPRoutes() {
-	out, stderr, err := util.RunOVNNbctl(
-		"--format=csv", "--data=bare", "--no-heading", "--columns=_uuid,output_port", "find", "Logical_Router_Static_Route", "options={ecmp_symmetric_reply=\"true\"}")
+func (oc *Controller) cleanExGwECMPRoutes() {
+	start := time.Now()
+	defer func() {
+		klog.Infof("Syncing exgw routes took %v", time.Since(start))
+	}()
+
+	// Get all ECMP routes in OVN and build cache
+	ovnRouteCache := buildOVNECMPCache()
+
+	if len(ovnRouteCache) == 0 {
+		// nothing in OVN, so no reason to search for stale routes
+		return
+	}
+
+	// Build cache of expected routes in the cluster
+	// map[podIP][]nextHops
+	clusterRouteCache := make(map[string][]string)
+
+	// Find all pods serving as exgw
+	oc.buildClusterECMPCacheFromPods(clusterRouteCache)
+
+	// Get all namespaces with exgw routes specified
+	oc.buildClusterECMPCacheFromNamespaces(clusterRouteCache)
+
+	// compare caches and see if OVN routes are stale
+	for podIP, ovnRoutes := range ovnRouteCache {
+		// pod IP does not exist in the cluster
+		// remove route and any hybrid policy
+		if _, ok := clusterRouteCache[podIP]; !ok {
+			continue
+		}
+
+		// podIP exists, check if route matches
+		expectedNexthops := clusterRouteCache[podIP]
+		for _, ovnRoute := range ovnRoutes {
+			// if length of the output port is 0, this is a legacy route (we now always specify output interface)
+			if len(ovnRoute.outport) == 0 {
+				continue
+			}
+
+			node := util.GetWorkerFromGatewayRouter(ovnRoute.router)
+			// prefix will signify secondary exgw bridge, or empty if normal setup
+			// have to determine if a node changed while master was down and if the route swapped from
+			// the default bridge to a new secondary bridge (or vice versa)
+			prefix, err := oc.extSwitchPrefix(node)
+			if err != nil {
+				// we shouldn't continue in this case, because we cant be sure this is a route we want to remove
+				klog.Errorf("Cannot sync exgw route: %+v, unable to determine exgw switch prefix: %v",
+					ovnRoute, err)
+			} else if (prefix != "" && !strings.Contains(ovnRoute.outport, prefix)) ||
+				(prefix == "" && strings.Contains(ovnRoute.outport, types.EgressGWSwitchPrefix)) {
+				continue
+			}
+
+			for _, clusterNexthop := range expectedNexthops {
+				if ovnRoute.nextHop == clusterNexthop {
+					ovnRoute.shouldExist = true
+				}
+			}
+		}
+	}
+
+	klog.Infof("OVN ECMP route cache is: %+v", ovnRouteCache)
```

Comment:
Ultimately this looks like:
```
I0810 09:38:00.828832      37 egressgw.go:713] OVN ECMP route cache is: map[10.244.1.5:[0xc000424a50 0xc000424d70] 10.244.2.6:[0xc000424c30 0xc000424cd0] 10.244.2.7:[0xc000424af0 0xc000424b90]]
```
Not really readable, so maybe just print the length of values for a particular key? This on a cluster with lots of pods will look a bit bad.